From 1309022adace863eed15544251cfdff73b018452 Mon Sep 17 00:00:00 2001
From: Loc Vu <loc.vu.zn@renesas.com>
Date: Mon, 20 Apr 2020 14:30:43 +0700
Subject: [PATCH 092/433] tmio: mmc: fix hardware interrupt timeout (CMD12)

Below issue occurred with eMMC on iwg23s:
"sh_mobile_sdhi ee300000.mmc: timeout waiting for hardware interrupt
(CMD12)"

This issue involve power management of mmc device.
Before setting clock for mmc, should resume callback for the device by
adding function pm_runtime_get_sync.

After stopping clock for mmc, should request idle for the device by
adding function pm_runtime_put.

Signed-off-by: hienhuynh <hien.huynh.px@rvc.renenas.com>
Signed-off-by: Quynh Nguyen <quynh.nguyen.xb@renesas.com>
---
 drivers/mmc/host/tmio_mmc_core.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/mmc/host/tmio_mmc_core.c b/drivers/mmc/host/tmio_mmc_core.c
index ed49a76..c9422dc 100644
--- a/drivers/mmc/host/tmio_mmc_core.c
+++ b/drivers/mmc/host/tmio_mmc_core.c
@@ -1066,13 +1066,16 @@ static void tmio_mmc_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)
 	case MMC_POWER_OFF:
 		tmio_mmc_power_off(host);
 		tmio_mmc_clk_stop(host);
+		pm_runtime_put(dev);
 		break;
 	case MMC_POWER_UP:
+		pm_runtime_get_sync(dev);
 		tmio_mmc_power_on(host, ios->vdd);
 		tmio_mmc_set_clock(host, ios->clock);
 		tmio_mmc_set_bus_width(host, ios->bus_width);
 		break;
 	case MMC_POWER_ON:
+		pm_runtime_get_sync(dev);
 		tmio_mmc_set_clock(host, ios->clock);
 		tmio_mmc_set_bus_width(host, ios->bus_width);
 		break;
-- 
2.7.4

